src/cargo/core/compiler/mod.rs

```diff
@@ -393,6 +393,22 @@ fn link_targets<'a, 'cfg>(
         .collect();
     let json_messages = bcx.build_config.json_messages();
 
+    // Find the executable for the current target (if any).
+    let mut executable : Option<PathBuf> = Option::None;
+    for output in cx.outputs(unit)?.iter() {
+        if output.flavor == FileFlavor::DebugInfo {
+            continue;
+        }
+        let bindst = match output.hardlink {
+            Some(ref link_dst) => link_dst,
+            None => &output.path,
+        };
```

Comment:
This logic is duplicated when we populate `Compilation::binaries`. Let's extract a helper function `fn exetutable_dest(unit: &Unit)`?

Reply:
It might need access to `cx` as well.

Reply:
@matklad Do you mean the `let bindst` logic?

Reply:
Yep! The "is this unit executable" logic needs to be different.

Reply:
@matklad Question: should the logic that populates `Compilation::binaries` be updated to include the `|| (unit.mode.is_any_test() && !unit.mode.is_check())` criteria as well?